### ISSUE TITLE
Setting Material.blending to "CustomBlending" no longer works blends as expected

### ISSUE THREAD

**Issue by gkjohnson** [mrdoob/three.js#23408]
**Describe the bug**

As of r137 using "CustomBlending" does not work as expected.

**To Reproduce**

Create a material that uses "CustomBlending" and a partial opacity. In r136 it is rendered as partially transparent while in r137 it is not.

***Code***

```js
const material = new THREE.MeshBasicMaterial( {
  blending: CustomBlending,
  opacity: 0.5
} );
```

***Live example***

* https://jsfiddle.net/c9v0jdrL/

**Expected behavior**

When using custom blending it renders as partially transparent as it did in r136.

**Screenshots**

From the js fiddle above. Left is a regular transparent cube. Right is using CustomBlending.

| r136 | r137 |
|---|---|
| ![image](https://user-images.githubusercontent.com/734200/152075404-386c3037-edb8-4967-a47d-b2db293abbe9.png) | ![image](https://user-images.githubusercontent.com/734200/152075371-cd9e048c-05c1-43ae-be5b-f55378cfca39.png) |

**Platform:**

 - Device: Desktop
 - OS: MacOS
 - Browser: Chrome
 - Three.js version: r137.5

**Comment by mrdoob** [comment]
See #23361

Does this work for you?

```js
const material = new THREE.MeshBasicMaterial( {
  blending: CustomBlending,
  opacity: 0.5,
  transparent: true
} );
```

**Comment by gkjohnson** [comment]
Using `transparent: true` does not work for my use case, no, and that's the reason I'm using CustomBlending in the first place. I'm using custom blending with the use case described [here](https://github.com/mrdoob/three.js/pull/18176#issuecomment-571272611) in the Gropu.renderOrder update thread (which I'd still like to move forward at some point 😁)

Basically I'm tinting the area where two objects overlap using the stencil buffer operations so draw order is very important:

1. draw object to be intersected and tinted
2. draw intersection volume invisibly with colorMask = 0 and stencil operations so only intersected volume is written to stencil.
3. draw an object with custom blending only where stencil buffer is set to tint the area.
4. repeat steps 2 and 3 for every intersection tint.
5. draw objects that are unaffected by the tint.

This jsfiddle from the above issue demonstrates the effect:

https://jsfiddle.net/sbygno54/1/

![image](https://user-images.githubusercontent.com/734200/152081581-7686353c-b50c-4c17-80e5-11a10537b056.png)


The white plane and sphere the environment being checked against, the green sphere is intersection volume being tested, the red area is the intersected area highlighted, and the square is the final object drawn without tint.

**Comment by mrdoob** [comment]
Hmm, isn't there any other way of achieving the same thing? I wouldn't want to revert #23361.

**Comment by mrdoob** [comment]
Maybe using `MultiplyBlending` instead? https://jsfiddle.net/h4jtkvo9/

**Comment by WestLangley** [comment]
@gkjohnson You are using custom blending without specifying the material properties that define the custom blending formula?

**Comment by gkjohnson** [comment]
> @gkjohnson You are using custom blending without specifying the material properties that define the custom blending formula?

The default values for `blendDst` and `blendSrc` for are good for this use case

> Hmm, isn't there any other way of achieving the same thing?

I don't believe there is.

> Maybe using MultiplyBlending instead? 

Multiply blending wouldn't work on dark colors -- additive blending works with any arbitrary environment.

> I wouldn't want to revert #23361.

TBH it seems like a pretty odd graphics feature to remove. At first glance I'd think that setting `alphaWrite` to false by default would be a better solution and allow users to fall back to the behavior they want.

I'm not familiar with the history of #23361 but from what I gather it looks like the "regression" mentioned is that when the canvas is to alpha: true and partially opaque (not marked as transparent) objects are drawn then the alpha value is written to the canvas causing the browser to blend it with the background contents is that right?

If that's the case why not just use zero out the contribution to the final canvas alpha using blending operations if transparent === false? Using v0.137.0 where the regression is it looks like these settings work:

```js
const material = new THREE.MeshNormalMaterial();
material.opacity = 0.25;
material.blending = THREE.CustomBlending;
material.blendDstAlpha = THREE.OneFactor;
material.blendSrcAlpha = THREE.ZeroFactor;
```

Here's a table of results with different blending settings from that fiddle:

| Mode | Snippet | Screenshot |
|---|---|---|
| NormalBlending | ![image](https://user-images.githubusercontent.com/734200/152084864-d1b427fd-d432-4d91-9a7e-014980ce69cf.png) |  ![image](https://user-images.githubusercontent.com/734200/152084743-58084fcd-bf4a-4265-94e3-d3e34b705bf6.png) |
| AdditiveBlending | ![image](https://user-images.githubusercontent.com/734200/152084924-bab0e853-3ab8-4d99-9728-ff012f07d933.png) | ![image](https://user-images.githubusercontent.com/734200/152084650-fdf1a526-1bae-446d-826a-f2aa1fe9a1ad.png) |
| CustomBlending | ![image](https://user-images.githubusercontent.com/734200/152085161-b05d0847-69ea-4b41-a163-3938b0860ff2.png) | ![image](https://user-images.githubusercontent.com/734200/152085176-d593fcf0-e741-4946-bf43-816a201a397a.png) |
| transparent = true | ![image](https://user-images.githubusercontent.com/734200/152084996-68e6aafc-e525-45e7-b0e8-6966e5012f56.png) | ![image](https://user-images.githubusercontent.com/734200/152084631-924ac605-7c6b-4efc-8161-5a56e878cda1.png) |

**Comment by mrdoob** [comment]
> I'm not familiar with the history of #23361 but from what I gather it looks like the "regression" mentioned is that when the canvas is to alpha: true and partially opaque (not marked as transparent) objects are drawn then the alpha value is written to the canvas causing the browser to blend it with the background contents is that right?

That right.

> If that's the case why not just use zero out the contribution to the final canvas alpha using blending operations if transparent === false? 

I think that should be a good solution too. Would you like to do a PR with it?

We'll then have to give another go at `alphaWrite`/`alphaMode` for glTF compatibility.

**Comment by mrdoob** [comment]
Actually...

> | Mode | Snippet | Screenshot |
> |---|---|---|
> | NormalBlending | ![image](https://user-images.githubusercontent.com/734200/152084864-d1b427fd-d432-4d91-9a7e-014980ce69cf.png) |  ![image](https://user-images.githubusercontent.com/734200/152084743-58084fcd-bf4a-4265-94e3-d3e34b705bf6.png) |

That one is wrong. In this case the cube shouldn't have that white tinting.

**Comment by WestLangley** [comment]
>@gkjohnson You are using custom blending without specifying the material properties that define the custom blending formula?

>>The default values for blendDst and blendSrc for are good for this use case.

Hmmm... I would have expected the default values for `CustomBlending` match the values that produce `NormalBlending`. At least doing so would have been a reasonable default.

**Comment by gkjohnson** [comment]
> That one is wrong. In this case the cube shouldn't have that white tinting.

This is the normal case that's demonstrating the regression that was fixed in #23361. The other table rows are showing workarounds and behavior with different blend modes that work more as expected. With the fix I'm proposing the NormalBlending case will behave like the CustomBlending case with the equation adjustments I posted.

Perhaps there are better ways to handle this that make background / browser blending work more intuitively but I'd have to look into it more.

> Hmmm... I would have expected the default values for CustomBlending match the values that produce NormalBlending. At least doing so would have been a reasonable default.

It did previously and it still does when `.transparent === true` but this seems to be a new blending quirk of defaulting the canvas to alpha = true.

**Comment by WestLangley** [comment]
>Hmmm... I would have expected the default values for CustomBlending match the values that produce NormalBlending. At least doing so would have been a reasonable default.
>>It did previously...

Well, if that were the case, then you would have had no reason to use the default `CustomBlending` because it would have been the same as `NormalBlending`. Right?

**Comment by gkjohnson** [comment]
Yes you're right. I don't remember all the circumstances that led me to using CustomBlending as I wrote that code years ago other than just changing .opacity with transparent = false was somehow unreliable or not working as expected.

**Comment by mrdoob** [comment]
So you can achieve the same effect using CustomBlending then? 🤔

**Comment by gkjohnson** [comment]
> So you can achieve the same effect using CustomBlending then? 🤔

Can you rephrase the question? I'm not following. CustomBlending is currently what's broken and I want to fix.

Also do you recall what change specifically improved example performance from 40-50fps to 60fps from this comment? Was it just enabling alpha on the webglrenderer?

https://github.com/mrdoob/three.js/pull/23230#issuecomment-1012705191

**Comment by mrdoob** [comment]
Basically, apparently many systems do not create a webgl context with `alpha: false` so the browser has to emulate it.

My main laptop (Intel MacBook Pro 2016) is one of the worst cases apparently, so I've been dealing with subpar performance and noisy fans for years... Not it feels like a brand new laptop 😬

**Comment by mrdoob** [comment]
If you're able to figure out the right dst/src/dstAlpha/srcAlpha to use [here](https://github.com/mrdoob/three.js/blob/efbfc67edc7f65cfcc61a389ffc5fd43ea702bc6/src/renderers/webgl/WebGLState.js#L741-L743) that avoids having to do `gl_FragColor.a = 1.0` in the shader that'd be great. @Mugen87 and I tried but we couldn't figure it out.

**Comment by gkjohnson** [comment]
I don't think the `alpha: true` flag change should be reverted -- just that a different solution be used to address this artifact pointed out above.

> Setting transparent to false essentially means to me "alpha blending is disabled".

This is an unnecessarily heavy handed approach, I think. At the least blending could still be applicable when `blending === CustomBlending` so the capability isn't completely neutered. But I'm going to look at a couple options later today when I get a chance.

**Comment by gkjohnson** [comment]
PR here: #23416

> If you're able to figure out the right dst/src/dstAlpha/srcAlpha to use here that avoids having to do gl_FragColor.a = 1.0 in the shader that'd be great.

The blending modes could be set the same as when "transparent = true" for all objects in the opaque list but my understanding is that theres a performance impact when enabling blending. So if you don't need it so I think it's best to keep the OPAQUE flag there when custom blending is not enabled.
